File: 101-symmetric-tree/101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
		queue<TreeNode*> pending({ root->left, root->right });

		while (!pending.empty())
		{
			TreeNode* l = pending.front();
			pending.pop();
			TreeNode* r = pending.front();
			pending.pop();

			if (!l && r || l && !r) return false;
			if (l)
			{
				if (l->val != r->val) return false;
				pending.push(l->left);
				pending.push(r->right);
				pending.push(l->right);
				pending.push(r->left);
			}
		}

		return true;
    }
};

File: 101-symmetric-tree/101-symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return helper(root, root);
    }
    
    private boolean helper(TreeNode x, TreeNode y) {
        if (x == null && y == null)
            return true;
        if (x == null || y == null)
            return false;
        if (x.val != y.val)
            return false;
        return helper(x.left, y.right) && helper(x.right, y.left);
    }
}